New Claim Adjustment Reason Codes

gumption opened the pull request:
I was getting "is not a valid code for Adjustment Reason Code" errors, and discovered the codes in the 837I files I was processing were not in `map\codes.xml`, although they were listed in the latest update (11/1/2014) of the Claim Adjustment Reason Codes (http://www.wpc-edi.com/reference/codelists/healthcare/claim-adjustment-reason-codes/).

I added codes 205-268 and P1-P23.

I'm not sure if I succeeded in first resyncing with your changes to your master (to provide an initial binding to `cur_map` in `x12n_document.py`, rather than wrapping the `del cur_map` in a `try/except` clause). I followed the instructions here: https://help.github.com/articles/syncing-a-fork/

Testing your `cur_map` modification on the 837I files I was processing before does not generate any `UnboundLocalError` exceptions. Gaining approval to share the de-identified file will take some time (and may not be possible).

gumption commented:
Looks like there are also some codes missing from the `remark_codes` list in `map/codes.xml`:

http://www.wpc-edi.com/reference/codelists/healthcare/remittance-advice-remark-codes/

I won't update this file until / unless it's clear I successfully merged original branch ... and/or the pull request is accepted.

azoner commented:
Thanks for the added codes. In hindsight, it was a bad idea to validate frequently changing codes with sources outside the X12N spec.  I would prefer to remove those code validations.

It's fine if you are not able to supply test files.  My concern is that the cur_map variable will only be unset if no GS segment is found.  Could you check the output of x12valid --verbose and x12norm -e?  I want to ensure that the low level X12 parsing is working correctly for your files.

gumption commented:
Ah, I was just in the process of updating other code sets ... and belatedly realized I probably should be creating a new `<version>` element for each ... but if your preference is to remove them, then I will desist.

I will conduct the validation you requested tomorrow (I can only access the files via a secure, wired connection at the office).

azoner commented:
I didn't remember in how many places the adjustment reason codes were
referenced.  It probably would be best to leave the validation as it is.
So, any additional code updates would be welcome.

---------------------
John Holland
john@zoner.org

On Thu, Feb 12, 2015 at 8:25 PM, Joe McCarthy <notifications@github.com>
wrote:

> Ah, I was just in the process of updating other code sets ... and
> belatedly realized I probably should be creating a new <version> element
> for each ... but if your preference is to remove them, then I will desist.
>
> I will conduct the validation you requested tomorrow (I can only access
> the files via a secure, wired connection at the office).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/azoner/pyx12/pull/33#issuecomment-74189080>.
>

gumption commented:
When I run `x12valid --verbose` on one of my "good" 837I files, with 75 claims, it reports 31 instances of `SEG:3 - Mandatory loop "Claim Information" (2300) missing` errors (but no other errors), but a a usable XML file (with 75 parsed claims) is created. 

When I run `x12valid --verbose`  on one of the files that previously generated the `UnboundLocalError` exception for `del cur_map`, it reports many types and instances of errors (output is > 100K lines), and the output XML file has nothing but `ISA_LOOP`s.

When I run `x12norm -e` on a "good" claims file, and pipe it through `wc`, it generates output with the same number of lines and words, but a slightly smaller number of characters, as are in the source file. Since the # of characters differs by the # of lines, I interpret this as a stripping of `\r` characters (my source files have lines terminated with `\r\n` sequences).

   4958    6989  122332 (x12norm -e $file | wc -l)
   4958   6989 127290 (wc $file)

When I repeat this with one of the problematic files (for which `UnboundLocalError` exceptions were raised), I get the following:

   92423   64248 11645189 (x12norm -e $file | wc -l)
   46213    64248 11645676 (wc $file)

This is consistent with the extra line breaks that appear when I use `less $file`. 

I initially thought this may be due to `\r\n` sequences used for newlines, but removing `\r` did not have any effect on `x12valid`. 

Running `od -c $file` enabled me to see that there are extraneous blank spaces at the end of the lines (after the tilde, but before the newline character sequence) in the problematic files.

I ran `sed -e "s/~.*$/~/g" $file` to create new versions of the problematic files, and all seems well now.

If the trailing blank spaces were responsible for the parsing problems, I wonder if it would be worthwhile to modify the X12 parser to ignore trailing blank spaces (after the tilde), or, at least, to note them as errors, but remove the trailing whitespace and continue parsing.

gumption commented:
I've modified all code sets that have updated specifications at http://www.wpc-edi.com/reference/ with new `<version>` elements. 

I've also updated some of the error messages to be more consistent (data elements that were too short had messages starting with "Data element.."; those that were too long had messages starting with "Element...").

I will commit them to my fork and generate a new pull request once I've tested everything on my full set of claims files.

azoner commented:
That's exactly what I was worried about - the trailing whitespace was
causing the following GS segment to never be seen.  I can write an error
check for that.  I don't believe that should be considered valid X12.

x12norm --fix tries to correct some basic X12 structure problems.  I'd like
it to be able to handle and fix trailing whitespace.

Thanks for the work on the updated codesets. I believe the codeset version
is ignored.  Another thing that seemed like a good idea ten years ago but
was less easy to manage.

---------------------
John Holland
john@zoner.org

On Fri, Feb 13, 2015 at 11:04 AM, Joe McCarthy <notifications@github.com>
wrote:

> I've modified all code sets that have updated specifications at
> http://www.wpc-edi.com/reference/ with new <version> elements.
>
> I've also updated some of the error messages to be more consistent (data
> elements that were too short had messages starting with "Data element..";
> those that were too long had messages starting with "Element...").
>
> I will commit them to my fork and generate a new pull request once I've
> tested everything on my full set of claims files.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/azoner/pyx12/pull/33#issuecomment-74277006>.
>
